pipeline/subtraction.py

```diff
@@ -47,39 +128,61 @@ def run(self, *args, **kwargs):
-            #  I put in a really stupid workaround because for some reason
-            #  the reference FITS file and the exposure FITS file are not
-            #  aligned or shaped the same (there's a 90 degree rotation
-            #  and some extra rows and columns that need to be trimmed).
-            #  I have a feeling this will be solved when we insert the
-            #  alignment phase into the subtraction. See issue #128
-            sub_image.data = image.data[:100, :100] - ref.image.data[:100, :100]
-            if sub_image.provenance is None:
+
+            # manually add the upstreams of reference image's products (SourceList, PSF, WCS, ZP)
+            upstreams = prov.upstreams
+            upstreams.append(ref.sources.provenance)
+            upstreams.append(ref.psf.provenance)
+            upstreams.append(ref.wcs.provenance)
+            upstreams.append(ref.zp.provenance)
+            prov.upstreams = upstreams  # must re-assign to make sure list items are unique
+            prov.update_id()
+            prov = session.merge(prov)
+            sub_image = ds.get_subtraction(prov, session=session)
+
+            if sub_image is None:
+                self.has_recalculated = True
+                # use the latest image in the data store,
+                # or load using the provenance given in the
+                # data store's upstream_provs, or just use
+                # the most recent provenance for "preprocessing"
+                image = ds.get_image(session=session)
+                if image is None:
+                    raise ValueError(f'Cannot find an image corresponding to the datastore inputs: {ds.get_inputs()}')
+
+                sub_image = Image.from_ref_and_new(ref.image, image)
                 sub_image.provenance = prov
-            else:
-                if sub_image.provenance.id != prov.id:
-                    raise ValueError('Provenance mismatch for sub_image and provenance!')
+                sub_image.provenance_id = prov.id
+
+                # make sure to grab the correct aligned images
+                new_image = sub_image.aligned_images[sub_image.new_image_index]
+                ref_image = sub_image.aligned_images[sub_image.ref_image_index]
+
+                if self.pars.method.lower() == 'naive':
+                    outim, outwt, outfl = self._subtract_naive(new_image, ref_image)
+                elif self.pars.method.lower() == 'hotpants':
+                    raise NotImplementedError('Hotpants not implemented yet')
+                elif self.pars.method.lower() == 'zogy':
+                    outim, outwt, outfl, score, psf = self._subtract_zogy(new_image, ref_image)
+                else:
+                    raise ValueError(f'Unknown subtraction method {self.pars.method}')
+
+                sub_image.data = outim
+                sub_image.weight = outwt
+                sub_image.flags = outfl
+                if 'score' in locals():
```

Comment:
This kind of thing always scares me; if later we're editing code above in this function, we could well create a local variable named 'score'.  I'd feel better of the `self._subtract_*` methods returned a dictionary rather than an unnamed tuple, and then we could look for keys in that dictionary.

Reply:
good idea.